compiler/qsc_frontend/src/compile/preprocess.rs

```diff
@@ -193,3 +197,75 @@ fn matches_config(attrs: &[Box<Attr>], capabilities: TargetCapabilityFlags) -> b
     capabilities.contains(found_capabilities)
         && (disallowed_capabilities.is_empty() || !capabilities.contains(disallowed_capabilities))
 }
+
+// Visitor to remove spans from circuit callables defined in QSC files.
+// This will remove the spans for the contents of these circuit callables,
+// but it is important for the language server that the span for the callable itself
+// is preserved, so that the user can navigate to the definition of the callable.
+pub(crate) struct RemoveCircuitSpans {
+    qsc_spans: Vec<Span>,
+    inside_stmt: bool,
+}
+
+impl RemoveCircuitSpans {
+    pub(crate) fn new(sources: &SourceMap) -> Self {
+        let qsc_spans = sources
+            .iter()
+            .filter(|source| {
+                std::path::Path::new(source.name.as_ref())
+                    .extension()
+                    .is_some_and(|ext| ext.eq_ignore_ascii_case("qsc"))
+            })
+            .map(|source| {
+                let start = source.offset;
+                let end = start
+                    + u32::try_from(source.contents.len()).expect("source length exceeds u32::MAX");
+                Span { lo: start, hi: end }
+            })
+            .collect();
+
+        Self {
+            qsc_spans,
+            inside_stmt: false,
+        }
+    }
+}
+
+impl MutVisitor for RemoveCircuitSpans {
+    fn visit_package(&mut self, package: &mut Package) {
+        // We only want to visit namespaces
+        package.nodes.iter_mut().for_each(|n| match n {
+            TopLevelNode::Namespace(ns) => self.visit_namespace(ns),
+            TopLevelNode::Stmt(_) => {}
+        });
+    }
+
+    fn visit_namespace(&mut self, namespace: &mut Namespace) {
+        // We only want to visit circuit callables
+        namespace.items.iter_mut().for_each(|item| {
+            if let ItemKind::Callable(callable) = item.kind.as_mut() {
+                // Check if the callable's span is inside a QSC file's span
+                self.qsc_spans
+                    .iter()
+                    .any(|s| s.lo <= callable.span.lo && s.hi >= callable.span.hi)
+                    .then(|| {
+                        self.visit_callable_decl(callable);
+                    });
+            }
+        });
+    }
+
+    fn visit_stmt(&mut self, stmt: &mut Stmt) {
```

Comment:
This logic isn't generally safe and is making assumptions about the AST which may or may not be safe depending on the qsc compilatation of the circuits.

Reply:
Can you describe more what you mean when you say it isn't safe? Perhaps you are referring to the fact that there are statements that we don't want to remove spans from, and this relies on the other visit functions to make sure this only gets called for circuit operations.

Reply:
I see what you mean now. The logic in this function shouldn't be setting `inside_stmt` to `true` and `false`. It should be saving the value as it is, setting it to true, then afterwards restoring the origin value. This avoids the issues of when a statement existing inside another statement.

Reply:
Though this might be unnecessary if all we want to do is de-span statements and not their underlying expressions.

Reply:
Decided to only de-span statements.